### ISSUE TITLE
-fod from tckglobal

### ISSUE THREAD

**Issue by thijsdhollander** [MRtrix3/mrtrix3#478]
Since the changes in commit c45c33fbf811fbd7d9374848f24912d23136689b , I've been getting questions from users, where their initial formulation suggest serious confusion between the output from `tckglobal` using the `-fod` option and `msdwi2fod` .  The question on the community forum at http://community.mrtrix.org/t/dark-blocks-in-fod-output-from-tckglobal/126 is the latest occurrence of such confusion.  This is getting quite annoying, and at times starts to reflect badly on the MSMT-CSD method: people often seem to think that `tckglobal` first effectively does MSMT-CSD and then runs a global tractography on the output of that; and hope to get this output optionally via the `-fod` option.  After explaining what `tckglobal` actually does, and what the `-fod` option outputs, I've had someone actually asking me "why don't you call it a TOD then?".  If the `tckglobal` result is not that good (due to various possibilities; users might struggle with the tractography modelling parameters and assumptions for their data), some people are concluding that MSMT-CSD had not done a good job, and that they can see that that was the problem via the `-fod` output.

I know we've been referring to it as an FOD in the paper; but I'm starting to seriously question the confusion that this is introducing.

Now let's simply be down to earth, open and honest about this: @dchristiaens , I understand you want to avoid having to include a reference to the TOD paper --all good with me (to leave you with that choice)-- but we should at least consider renaming the `-fod` option to `-pod` or something then (Particle OD).  In the end, it's about what the code effectively _implements_; it's even using apodised delta functions to provide that particle _mapping_ per default.  That's not even the mapping you use to fit the particles to the data throughout the algorithm.  It _is_ effectively a post-processing step; it _is_ a mapping of the output of `tckglobal`, that inherently embodies all the tractography modelling assumptions and parameter choices of it.  And as you mentioned on the discourse, it's even only concentrating the whole particle contribution at the centre point.  Calling that an FOD for anything beyond a particle length of 1mm... I don't know, it just doesn't feel right to me.  I feel this should be reconsidered before the release.  It's just not responsible to cause confusion about one of our most fundamental terms: the FOD, of which we also associate the amplitude with other quite central concepts.  I mean this well, and I'm all for avoiding such crucial confusion.

**Comment by bjeurissen** [comment]
Hi, Thijs.

I can see how users will easily get confused and I agree that it should be avoided. I am in favor of your suggestions.

Also, I think in general we should not be trying to avoid eachother's references but rather try to include them if relevant. We should not compete for references but rather facilitate them for each other. Will benefit us as a team in the long run.

**Comment by jdtournier** [comment]
@bjeurissen, I'm guessing you didn't mean to close this issue...? It seems there's still quite a bit to resolve... 

It seems to me that this `-fod` option would be better renamed to `-predicted_fod` - unless I'm getting the wrong end of the stick?

I'd also like to figure out why users would even think of using `tckglobal` to do MSMT analysis. I haven't had a look at the docs in a while, it'll be worth having a good critical look at the current docs from a user's point of view, see whether there's anything that might have given the wrong impression or something...

**Comment by bjeurissen** [comment]
@jdtournier, closed it by accident, sorry... Was using a smartphone and somehow hit the wrong button.

**Comment by thijsdhollander** [comment]
It's basically users looking for "something to do with their multi-shell data".  At the moment, options are sparse, so they quickly spot the MSMT-CSD and the global tracking if they look a bit further.  The mistake they make is that they reason that global tracking runs from MSMT-CSD results (but for implementation reasons, it's in a single command).  One reasoning I've picked up is that that seems equivalent to tensor tracking via `tckgen` (where we also require the raw data to be provided, rather than a tensor image).  But they all even only started thinking in these ways, because they spotted the `-fod` option; and from there on just reason in a classic manner of "model fitting first, then tracking".  We've had the FOD since the very beginning of CSD; so they only recognise it as a voxel-wise (independently fitted) model.  A distribution of tracks, on the other hand, is recognised as a *track* (orientation) distribution.  Tracks of course are always intended to fit (parts of) the data.  Even probabilistic tracking does that, by sampling the FOD; but just doesn't get the density right across voxels.  Seeding in a more clever way can help that.  SIFT can help even more; modifying the track distribution to even better fit the data.  But it's still a track distribution; obeying certain tracking constraints, regularizations and most importantly: assumptions.

The mistake would be less likely to happen if users would actually read the full global tracking paper; but being pragmatic, we've got to realise that that is not always what the less technically inclined users tend to do (especially not in case of a highly technical paper).  We could blame them; but still: why even cause confusion, if confusion can be easily avoided by changing the term to reflect the difference?  The term "FOD" (in MRtrix) has had a very specific meaning and conotation since the introduction of CSD.  It's easy to see how users think of it that way.  I don't think they intend to use that command for MSMT-CSD analysis per se; but they still think it runs off of it (especially since they've now been included in the "new version" (merge) at the same time), and use the `-fod` option to check if the "cause" of undesirable tracking results is with the MSMT-CSD method somehow.  Clearly, this may reflect badly on it.

Therefore, I'd steer away from anything (containing) "FOD" in this context.  There's plenty of other reasonable namings possible (the original version had "TOD", the commit that changed this into `fod` even states the difference is mute; so why not keep it clear then?).

About the reference, it's a different discussion, but: I share @bjeurissen 's opinion here.  I've even requested it ( https://github.com/MRtrix3/mrtrix3/pull/415#issuecomment-163720082 ), since the command effectively *implements* about every formula in the paper to provide the mapping (we often include references for much less).  Opinions may differ, so I'm good with leaving the choice with the author; but I'm myself also all for an inclusive approach in general.  I don't think *including* a reference of the type "when using the `-...` option" would somehow lower the value of the "main reference".

**Comment by dchristiaens** [comment]
The way I see it, the fODF is the orientation distribution and density of white matter fibres in the brain, period. Many methods aim to estimate this distribution from DWI data, some better than others, but none can claim to produce anything other than an estimate. In global tractography, we aim to reconstruct the full tractogram that best explains the data in a single optimization process. As opposed to CSD, the model representation is not discretised on a voxel grid, but rather as a set of particles (line segments), each of which has a fixed and equal contribution to the signal. The mechanism by which this kernel is reoriented along the direction of each particle employs Dirac delta functions, as eloquently illustrated in Donald’s original SD paper (Tournier et al., 2004, Fig. 1). This leads to an intermediate "particle ODF” representation (I've used that name in chats with Thijs before) that is quantitatively related to the data. Therefore, under the assumptions of SD, this "particle ODF" is a reasonable estimate of the true fibre ODF. Sure, the spatial prior in global tractography will affect this estimate &mdash; that is by design. Think for example of noisy data where such spatial priors may help clean up spurious CSD peaks. This is not new or unique to global tractography: many spatially-regularized CSD methods have been introduced, including one by Donald, all of which use the same term fODF. For these reasons I am comfortable naming this intermediate representation _a_ fibre ODF.

Of course, the mechanism by which it is constructed shares similarities with Thijs’ TOD representation, and this is the main reason for why I went back and forth between both names for a while. However, because a TOD can represent any tractogram it is inherently decoupled from the data. As you’re well aware, the TOD representation of a streamline tractogram has no quantitative relation to the DWI data whatsoever. Therefore, we finally decided on the name "fibre ODF” to stress the relation of global tractography to the data and to highlight the difference with streamline tracking. We, the authors of the paper, includes Thijs.

That said, if there would be any kind of confusion about the difference with MT-CSD, I fully agree that it should be clarified. Have a look at the latest [user guide](https://github.com/MRtrix3/mrtrix3/blob/update-tckglobal-docs/docs/workflows/global_tractography.rst) and let me know if you find a possible source of confusion. I’m also happy with Donald’s suggestion to rename the `-fod` output option to `-predicted_fod` if it clears things out, although I would prefer `-intermediate_fod` or `-ancillary_fod`.

Finally, I don’t think our users are like blind monkeys that somehow expect `tckglobal` to do CSD (it’s called `tck*` for a reason). Even if some of them are, they’d be quickly eaten by the three-headed dragon that global tractography can be. The people I spoke with, at least, had all read the paper and had a fairly good understanding of what it intends to do. I don’t think that forum thread Thijs refers to had much to do with such misunderstanding either. Again, if there’s any kind of confusion, it needs to be addressed; but I wouldn’t want to discourage anyone from comparing different methods either.

P.S.: Concerning the convolution with apodised PSFs, first introduced in Dave’s reorientation paper, they just make visualisation so much nicer. I incorporated them only recently to simplify things for users, but if people feel they may be causing confusion, they can go. It will be a few lines of code less, and a few additional lines in the user docs calling `shconv`.

P.<sup>2</sup>S.: Regarding citations, @bjeurissen said _"I think in general we should not be trying to avoid eachother's references but rather try to include them if relevant. We should not compete for references but rather facilitate them for each other. Will benefit us as a team in the long run.”_ I couldn’t agree more. It is also exactly what I did in the paper and in the [user guide](https://github.com/MRtrix3/mrtrix3/blob/update-tckglobal-docs/docs/workflows/global_tractography.rst). In anyone feels some are missing, by all means let me know. When it comes to the `tckglobal -help` docs, I do prefer to keep things short and to the point. In fact, I simply adopt the approach taken by many, if not all other commands: provide the relevant references to the implemented method, not to objects it produces, alternatives, or vaguely related work. A few examples: `msdwi2fod` only references Jeurissen et al. (2014), not Tournier et al. (2004/2007); `dwi2tensor` doesn’t consider a reference to Basser et al. (1994) might be relevant; and `fod2dec` doesn’t reference Pajevic and Pierpaoli (1999) either. If people feel we need to switch policy on that front, I would happily include references to Reisert et al. (2011) and Jeurissen et al. (2014), the two cornerstones of our global tractography work.

**Comment by jdtournier** [comment]
OK, so this is a case of TOD meets FOD... 

The problem is that global tractography methods (including SIFT and related methods) strive to ensure the TOD approximates the FOD as closely as possible.  By that rationale, we could argue that the TOD derived from a SIFTed tractogram can be labeled as an FOD - I'm not sure I'm comfortable with that. I appreciate the dilemma here, but I don't think it can be viewed as a FOD in the way that a voxel-wise estimate might, given the massive contribution of the tractography assumptions and priors - although I agree that the lines get blurred when spatial regularisers are used in voxel-wise measures. 

Which is why my preference is for the _predicted FOD_ - I'm not sure I get the rationale behind 'ancillary' or 'intermediate', but that's a minor point.

But all this is I think a distraction from the bigger issue. I don't think renaming it will fix @thijsdhollander's original concerns with user confusion particularly. If there is confusion, it needs to be addressed squarely in the documentation. But then, @dchristiaens suggests that actually, there is no confusion - those users who have asked about it seem to have been perfectly aware of what they were doing. Is that right? In which case, it strikes me that there isn't much of a problem... (?)

As to the 'cite or don't cite' question, again I think this is a red herring. I don't think anyone in this team would think twice about giving credit where it is due. In this particular case, I'd have to say that even if the option were renamed to `-tod`, it would be a bit of a stretch to therefore include this reference in the command inline documentation - it is clearly very peripheral to the method as implemented. If the resulting tractogram were subsequently used to generate a TOD, then that command should cite the paper (I assume it does?). If however this is a problem, it may be better/cleaner to remove the option altogether, as @dchristians suggests, since it isn't actually used in the algorithm as such - it's just provided for convenience. Assuming that the output of `tckmap -tod` is equivalent, then maybe that's where the functionality should live...?

**Comment by dchristiaens** [comment]
> But all this is I think a distraction from the bigger issue. I don't think renaming it will fix @thijsdhollander's original concerns with user confusion particularly. If there is confusion, it needs to be addressed squarely in the documentation. But then, @dchristiaens suggests that actually, there is no confusion - those users who have asked about it seem to have been perfectly aware of what they were doing. Is that right? In which case, it strikes me that there isn't much of a problem... (?)

That is my impression, yes. I get questions mostly regarding connectomics or regarding tractography in “low quality” data that might benefit from a global approach. Personally, I have not seen anyone somehow confusing `tckglobal` and `msdwi2fod`. Admittedly, that’s only about 10 users (< 1% of the mrtrix user base?), but that might just mean there’s even less of an issue here (?): there doesn’t seem to be that many people using it. But like I said, if the docs are unclear I’m happy to clarify things. I have a fair bit of work on the user guide when it comes to explaining different parameters anyhow.

> If the resulting tractogram were subsequently used to generate a TOD, then that command should cite the paper (I assume it does?). If however this is a problem, it may be better/cleaner to remove the option altogether, as @dchristians suggests, since it isn't actually used in the algorithm as such - it's just provided for convenience. Assuming that the output of `tckmap -tod` is equivalent, then maybe that's where the functionality should live...?

I’m not sure I fully understand what you mean. Just a brief clarification to be sure we’re on the same page:
![Fig. 1 from the paper](https://cloud.githubusercontent.com/assets/7466983/13905748/04d34c8c-eec9-11e5-885e-38e1ea9a6459.jpeg)
At any point throughout the algorithm, the DWI signal is simulated from the current particle configuration, by mapping all particles located in the voxel (a) to delta functions (b) and adding these to obtain the _intermediate_ ODF I was talking about. Spherical convolution of this ODF with the WM response then simulates the WM signal that is compared with the actual data. This all happens _online_: the ODF (c) is stored in memory (as a 4D image) throughout the entire optimization, and updated locally upon each change to the particle configuration, in each iteration. It is therefore inherent to the algorithm, and not a postprocessing step in any way. The only postprocessing currently involved, is a convolution of this intermediate ODF representation (c) with an apodisation kernel to produce the output `-fod`.

For parameter tuning and debugging, it is vital to have access to the _sum-of-delta_ ODF as shown in (c), and I wish to keep an option to export it. What is not essential, and purely intended for visualisation, is to export the _sum-of-aPSFs_ fODF. So what I was suggesting is to remove the apodisation step &mdash; assuming its output may somehow be confusing. We could then also rename the output option `-fod` to a name that better reflects that it's part of the optimization. In the paper, we called it an _ancillary_ result of the global tractography approach, hence my suggestion.

**Comment by thijsdhollander** [comment]
> The problem is that global tractography methods (including SIFT and related methods) strive to ensure the TOD approximates the FOD as closely as possible. By that rationale, we could argue that the TOD derived from a SIFTed tractogram can be labeled as an FOD - I'm not sure I'm comfortable with that. I appreciate the dilemma here, but I don't think it can be viewed as a FOD in the way that a voxel-wise estimate might, given the massive contribution of the tractography assumptions and priors - although I agree that the lines get blurred when spatial regularisers are used in voxel-wise measures. 

Fully agreed.

> Which is why my preference is for the predicted FOD - I'm not sure I get the rationale behind 'ancillary' or 'intermediate', but that's a minor point.

Still think the "F" in there may lead to confusion.  I think "ancillary" or "intermediate" (instead of "predicted") are not resolving the issue; again it may sound as if a voxel wise independent fit was done first (which would be MSMT-CSD in this case), and that is output as an "ancillary" or "intermediate" result.

*I'm not looking to pick fights here*; as I said up front: *I mean this well* and I'm honestly concerned about this, so let's not get personal.  We've discussed this ages ago as well (when writing it up), I have never been in favour of confusing users with "FOD" in this context, but left the final choice with you.  It's easy to see where we were coming from in the initial commit of the code.  I still have my own opinion, and am simply looking to express it out of genuine concern.  And in this case, it's about the context in which it's implemented: MRtrix, which has various CSD algorithms that produce the FOD that users are familiar with.

So let's look for a consensus, from the last two replies.  I'm happy with:

1. Renaming it "sum-of-delta ODF", as you suggest.  That makes clear *what it is*.  The option could be just named "-odf" (so it's not unwieldy), the description could clarify that it is a sum-of-delta ODF maintained during the execution of the algorithm.
2. Removing the option to output it convolved with an apodised PSF.  The "sum of delta ODF" remains available for debugging purposes.
3. Stating that it is *not* the TOD, so users are properly redirected to `tckmap` for all purposes of computing a proper TOD; the full TOD that the output tracks map to.  This command has the proper reference.

**Comment by jdtournier** [comment]
OK, clearly I need to read the paper properly first...

@dchristiaens: apologies, I'd assumed you'd never have needed to explicitly generate the FOD, since you can happily use the coefficients for the response directly, rather than those of a delta function... My mistake.

@thijsdhollander: I personally think 'predicted FOD' (or ancillary/intermediate if you really prefer that) is better than 'sum of delta functions' - the latter is completely obscure unless you've read the details of the paper, whereas the former ought to make sense to anybody who can appreciate that global tractography is an inverse problem. I prefer 'predicted' from that point of view because it fits in with my way of thinking: you have a model, you fit to the data, and see what predictions it provides. But that's just my preference. I can see @dchristiaens's point that it's not central here, and that actually the version produced by the `-fod` option is modified further by apodisation; nonetheless, I think it is best described as the FOD predicted by the model.

Coming to my earlier comment about how widespread this confusion is though: I'm still not sold on there being a problem here. The post @thijsdhollander` originally links to doesn't seem to indicate any confusion: it looks like the user was using the predicted FOD as a debugging tool, which is pretty much exactly what it's for, as far as I can tell. Maybe there's been other communications  through other channels that I haven't seen? 

Regardless, I really think the main thing here is to fix the documentation at a higher level. Maybe the problem is that we simply provide different applications with no overarching guidance as to what to use when and for what

**Comment by jdtournier** [comment]
Sorry - goofed again... Butter fingers on mobile device.

**Comment by jdtournier** [comment]
_Continued from above_

If we really have to modify the command itself, I think the simplest may be to remove the offending option, especially if it can be replicated with a `shconv` call (but that would need to be documented). But personally, I don't have a problem with it staying as-is, especially if the option description itself states clearly what it is, how it's produced, and what it's for (presumably tuning & debugging, mostly?). The option itself can be renamed to something else if you want, I'm not too bothered about that - I really don't see many users genuinely getting confused by the name, if there is confusion, it must originate elsewhere. Maybe the simplest thing might be to ask those users you're thinking might have been confused by this and ask them outright why they thought `tckglobal` might be the best way to get the FOD for multi-shell analysis? Then we could fix that - and if it turns out the `-fod` option was indeed the culprit, we can ask them whether the changes we have in mind would avoid said confusion. Would that be an option?

**Comment by thijsdhollander** [comment]
I can live with renaming to "-predicted_fod" and a proper description as I mentioned before, in the documentation of the option in the command.  I would appreciate removal of the apodised PSF convolved output; it's distracting from the debugging use, and suggesting again otherwise.  As mentioned, it's not people using the command for MSMT-CSD analysis; it's rather people using it to debug bad global tracking results, but believing that they were inherently generated of a normal MSMT-CSD; then seeing the bad result reflected in the `-fod` option, blaming MSMT-CSD for it, but not the tracking (as should've been in those cases where the outcome was not good).  The user on the forum also generated a true MSMT-CSD result for comparison of the encountered bug, stating surprise that they did not match.  So let me rephrase my proposal, to comply with the suggestion:

1. Renaming it `-predicted_fod`, as you suggest.  I am on the same front of preferring *"predicted"* to either "ancillary" or "intermediate", as the latter two may again potentially suggest that they are generated in an earlier ("MSMT-CSD") step.  The description should state more clearly that it is a structure maintained during the tracking algorithm.
2. Removing the option to output it convolved with an apodised PSF.  This suggests something too close a TOD, while it doesn't want to be one.  Mapping functionality (to a TOD) should then be reserved for the `tckmap` command, so users can properly compare the distribution of the final tracks (not particles) to other proper track distributions.
3. Stating that it is not the TOD, so users are properly redirected to `tckmap` for all purposes of computing a proper TOD; the full TOD that the output tracks map to. This command has the proper reference.
 
These steps could, in my opinion, fix the documentation of the command up to a reasonable level.  More clarification in the command's help text is of course always welcome.

Just aside; but *indeed*, as...

> @dchristiaens: apologies, I'd assumed you'd never have needed to explicitly generate the FOD, since you can happily use the coefficients for the response directly, rather than those of a delta function...

...it might be even better (and simply do away with *any* possible confusion) to let the option return the `-predicted_signal`.  Because that is what the method does: it fits a global model to the signal; it tries to predict the signal.
